File: Problems/Algorithms/566. Reshape the Matrix/ReshapeMatrix.kt
class Solution {
    fun matrixReshape(mat: Array<IntArray>, r: Int, c: Int): Array<IntArray> {
        val n = mat.size; val m = mat[0].size
        if (n * m != r * c) return mat
        
        val ans = Array(r) { IntArray(c) }
        var col = 0; var row = 0
        for (i in 0 until n) {
            for (j in 0 until m) {
                ans[row][col] = mat[i][j]
                col++
                if (col == c) {
                    col = 0
                    row++
                }
            }
        }
        
        return ans
    }
}

File: Problems/Algorithms/566. Reshape the Matrix/reshape_matrix.rs
impl Solution {
    pub fn matrix_reshape(mat: Vec<Vec<i32>>, r: i32, c: i32) -> Vec<Vec<i32>> {
        let (row, col) = (mat.len(), mat[0].len());
        if row * col != (r * c) as usize { return mat; }
        mat.concat().chunks(c as usize).map(|i| i.to_vec()).collect()
    }
}

